**About Book Crossing Dataset**<br>

This dataset has been compiled by Cai-Nicolas Ziegler in 2004, and it comprises of three tables for users, books and ratings. Explicit ratings are expressed on a scale from 1-10 (higher values denoting higher appreciation) and implicit rating is expressed by 0.

Reference: http://www2.informatik.uni-freiburg.de/~cziegler/BX/ 

**Objective**

This project entails building a Book Recommender System for users based on user-based and item-based collaborative filtering approaches.

#### Execute the below cell to load the datasets

In [48]:
import pandas as pd
import numpy as np

In [49]:
#Loading data
books = pd.read_csv("books/books.csv", sep=";", error_bad_lines=False, encoding="latin-1")
books.columns = ['ISBN', 'bookTitle', 'bookAuthor', 'yearOfPublication', 'publisher', 'imageUrlS', 'imageUrlM', 'imageUrlL']

users = pd.read_csv('books/users.csv', sep=';', error_bad_lines=False, encoding="latin-1")
users.columns = ['userID', 'Location', 'Age']

ratings = pd.read_csv('books/ratings.csv', sep=';', error_bad_lines=False, encoding="latin-1")
ratings.columns = ['userID', 'ISBN', 'bookRating']

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'
/var/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py

### Check no.of records and features given in each dataset

In [50]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
ISBN                 271360 non-null object
bookTitle            271360 non-null object
bookAuthor           271359 non-null object
yearOfPublication    271360 non-null object
publisher            271358 non-null object
imageUrlS            271360 non-null object
imageUrlM            271360 non-null object
imageUrlL            271357 non-null object
dtypes: object(8)
memory usage: 16.6+ MB


In [51]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
userID      278858 non-null int64
Location    278858 non-null object
Age         168096 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


In [52]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
userID        1149780 non-null int64
ISBN          1149780 non-null object
bookRating    1149780 non-null int64
dtypes: int64(2), object(1)
memory usage: 26.3+ MB


## Exploring books dataset

In [53]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  \
0    Mark P. O. Morford              2002     Oxford University Press   
1  Richard Bruce Wright              2001       HarperFlamingo Canada   
2          Carlo D'Este              1991             HarperPerennial   
3      Gina Bari Kolata              1999        Farrar Straus Giroux   
4       E. J. W. Barber              1999  W. W. Norton &amp; Company   

                                           imageUrlS  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlM  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlL  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

### Drop last three columns containing image URLs which will not be required for analysis

In [54]:
books.drop(books.columns[-3:],axis=1,inplace=True)

In [55]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  
0    Mark P. O. Morford              2002     Oxford University Press  
1  Richard Bruce Wright              2001       HarperFlamingo Canada  
2          Carlo D'Este              1991             HarperPerennial  
3      Gina Bari Kolata              1999        Farrar Straus Giroux  
4       E. J. W. Barber              1999  W. W. Norton &amp; Company

**yearOfPublication**

### Check unique values of yearOfPublication


In [56]:
## print(sorted(books['yearOfPublication'].unique()))

As it can be seen from above that there are some incorrect entries in this field. It looks like Publisher names 'DK Publishing Inc' and 'Gallimard' have been incorrectly loaded as yearOfPublication in dataset due to some errors in csv file.


Also some of the entries are strings and same years have been entered as numbers in some places. We will try to fix these things in the coming questions.

### Check the rows having 'DK Publishing Inc' as yearOfPublication

In [57]:
print (books['yearOfPublication'][books['yearOfPublication']=="DK Publishing Inc"])
print (books['yearOfPublication'][books['yearOfPublication']=="Gallimard"])

209538    DK Publishing Inc
221678    DK Publishing Inc
Name: yearOfPublication, dtype: object
220731    Gallimard
Name: yearOfPublication, dtype: object


### Drop the rows having `'DK Publishing Inc'` and `'Gallimard'` as `yearOfPublication`

In [58]:
books.drop([209538,221678,220731],axis=0,inplace=True)

### Change the datatype of yearOfPublication to 'int'

In [59]:
books["yearOfPublication"] = books["yearOfPublication"].astype('int64')

In [60]:
books.dtypes

ISBN                 object
bookTitle            object
bookAuthor           object
yearOfPublication     int64
publisher            object
dtype: object

### Drop NaNs in `'publisher'` column


In [61]:
books['publisher'] = books['publisher'].dropna()

## Exploring Users dataset

In [62]:
print(users.shape)
users.head()

(278858, 3)


userID                            Location   Age
0       1                  nyc, new york, usa   NaN
1       2           stockton, california, usa  18.0
2       3     moscow, yukon territory, russia   NaN
3       4           porto, v.n.gaia, portugal  17.0
4       5  farnborough, hants, united kingdom   NaN

In [63]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
userID      278858 non-null int64
Location    278858 non-null object
Age         168096 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB


### Get all unique values in ascending order for column `Age`

In [64]:
users.sort_values(by=["Age"]).iloc[:100,:] # limites ad output was large

userID                                       Location  Age
162239  162240                         buffalo, new york, usa  0.0
130115  130116         lower templestowe, victoria, australia  0.0
135953  135954                   powder springs, georgia, usa  0.0
207490  207491                        manassas, virginia, usa  0.0
189924  189925                 san francisco, california, usa  0.0
87639    87640                         fort worth, texas, usa  0.0
87756    87757                        san antonio, texas, usa  0.0
275284  275285                            watauga, texas, usa  0.0
98711    98712                    randallstown, maryland, usa  0.0
153369  153370        brentwood, western australia, australia  0.0
63338    63339                         athens, tennessee, usa  0.0
176613  176614                          ojai, california, usa  0.0
136082  136083                         chicago, illinois, usa  0.0
218806  218807                  modena, emilia romagna, italy  0.0
51848    51849                          phoenix, arizona, usa  0.0
87780    87781               mexico d.f., mexico d.f., mexico  0.0
167875  167876             sydney, new south wales, australia  0.0
226266  226267                         kelso, washington, usa  0.0
143877  143878        rio de janeiro, rheinland-pfalz, brazil  0.0
106023  106024                     winnipeg, manitoba, canada  0.0
176661  176662              deventer, overijssel, netherlands  0.0
189979  189980    greifswald, mecklenburg-vorpommern, germany  0.0
4041      4042                         san marcos, texas, usa  0.0
253348  253349                       vienna, \n/a\", austria"  0.0
195691  195692                 kingston, puerto rico, jamaica  0.0
167770  167771                       parede, lisboa, portugal  0.0
153188  153189                  medellin, antioquia, colombia  0.0
203330  203331                      singapore, n/a, singapore  0.0
218063  218064               oxford,, england, united kingdom  0.0
130197  130198                   simi valley, california, usa  0.0
203416  203417                        brooklyn, new york, usa  0.0
270814  270815             sydney, new south wales, australia  0.0
186458  186459                   odivelas, odivelas, portugal  0.0
168457  168458                       manila, n/a, philippines  0.0
86979    86980                       calgary, alberta, canada  0.0
144379  144380                camberwell, virginia, australia  0.0
61988    61989                     newtown, pennsylvania, usa  0.0
189615  189616                         oradea, bihor, romania  0.0
52888    52889                           ivrea, troino, italy  0.0
251794  251795                        new york, new york, usa  0.0
52795    52796                              sumner, iowa, usa  0.0
81137    81138    charlottetown, prince edward island, canada  0.0
87020    87021                       toronto, ontario, canada  0.0
98874    98875                london, england, united kingdom  0.0
270443  270444                london, england, united kingdom  0.0
107711  107712            aberfeldy, scotland, united kingdom  0.0
270274  270275                  mainz-kastel, hessen, germany  0.0
252197  252198           mihovljan, hrvatsko zagorje, croatia  0.0
8432      8433                       santiago, galicia, spain  0.0
243083  243084                                orem, utah, usa  0.0
175879  175880       bury st edmunds, england, united kingdom  0.0
80774    80775           altdorf, baden-wuerttemberg, germany  0.0
102052  102053                        zurich, zh, switzerland  0.0
30619    30620                      crookston, minnesota, usa  0.0
87462    87463                          jogja, n/a, indonesia  0.0
242716  242717                         lincoln, nebraska, usa  0.0
107835  107836                     meridian, mississippi, usa  0.0
101992  101993                greenville, south carolina, usa  0.0
119366  119367              petaling jaya, selangor, ma

Age column has some invalid entries like nan, 0 and very high values like 100 and above

### Values below 5 and above 90 do not make much sense for our book rating case...hence replace these by NaNs

In [65]:
users["Age"][(users["Age"]<5) | (users["Age"]>90)] = np.nan

/var/anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


### Replace null values in column `Age` with mean

In [66]:
from sklearn.impute import SimpleImputer
impute = SimpleImputer(missing_values=np.nan, strategy='mean')
users["Age"] = impute.fit_transform(np.reshape(users["Age"].values,(-1,1)))

### Change the datatype of `Age` to `int`

In [67]:
users["Age"] = users["Age"].astype("int64")

In [68]:
print(sorted(users.Age.unique()))

[5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90]


## Exploring the Ratings Dataset

### check the shape

In [69]:
ratings.shape

(1149780, 3)

In [70]:
n_users = users.shape[0]
n_books = books.shape[0]

In [71]:
ratings.head(5)

userID        ISBN  bookRating
0  276725  034545104X           0
1  276726  0155061224           5
2  276727  0446520802           0
3  276729  052165615X           3
4  276729  0521795028           6

### Ratings dataset should have books only which exist in our books dataset. Drop the remaining rows

In [72]:
rating_ISBN = ratings["ISBN"].unique() # Create a list of ISBNs that are in Rating database
books_ISBN = books["ISBN"].unique() # create a list of ISBNs that are in Books database
notInBooks = list(set(rating_ISBN) - set(books_ISBN)) # Find ISBNs in Rating that dows not exists in Books
print (len(rating_ISBN), len(books_ISBN), len(notInBooks))

340556 271357 70408


In [73]:
# Delete the one that does not exist in Books
ratings.drop(ratings["ISBN"][ratings["ISBN"].isin(notInBooks)].index,axis = 0,inplace=True)

### Ratings dataset should have ratings from users which exist in users dataset. Drop the remaining rows

In [74]:
ratings_userID = ratings["userID"].unique() # Get list of users from Rating Database
users_userID = users["userID"].unique() # Get list of users from User Database
notInUsers = list(set(ratings_userID) - set(users_userID)) # check users that exist in rating but not in user database
print (len(ratings_userID), len(users_userID), len(notInUsers))

92106 278858 0


All users that are in Rating DB is also in Users DB

### Consider only ratings from 1-10 and leave 0s in column `bookRating`

In [75]:
ratings.drop(ratings[ratings["bookRating"] == 0].index,axis =0, inplace = True)

In [76]:
ratings.bookRating.unique()

array([ 5,  3,  6,  7,  9,  8, 10,  1,  4,  2])

0's have been removed

### Find out which rating has been given highest number of times

In [77]:
ratings.groupby("bookRating").count().sort_values(by='userID',ascending=False)

userID   ISBN
bookRating               
8            91804  91804
10           71225  71225
7            66401  66401
9            60778  60778
5            45355  45355
6            31687  31687
4             7617   7617
3             5118   5118
2             2375   2375
1             1481   1481

Maximum rating is 8 - 91,804 times

### **Collaborative Filtering Based Recommendation Systems**

### For more accurate results only consider users who have rated atleast 100 books

In [78]:
from surprise import SVD
from surprise import dataset
from surprise import Reader
from surprise import evaluate, print_perf
from surprise.model_selection import train_test_split

### Generating ratings matrix from explicit ratings


#### Note: since NaNs cannot be handled by training algorithms, replace these by 0, which indicates absence of ratings

In [79]:
class MyDataset(dataset.DatasetAutoFolds):

    def __init__(self, df, reader):

        self.raw_ratings = [(uid, iid, r, None) for (uid, iid, r) in
                            zip(ratings['userID'], ratings['ISBN'], ratings['bookRating'])]
        self.reader=reader

In [80]:
reader = Reader(line_format='user item rating', rating_scale=(1, 10))
data = MyDataset(ratings, reader)

### Generate the predicted ratings using SVD with no.of singular values to be 50

In [81]:
trainset, testset = train_test_split(data, test_size=.15)

In [82]:
algo = SVD()
algo.fit(trainset)

In [83]:
# data.split(n_folds=3)
# algo = SVD()
# perf = evaluate(algo, data, measures=['RMSE', 'MAE'])
# print_perf(perf)

### Take a particular user_id

### Lets find the recommendations for user with id `2110`

#### Note: Execute the below cells to get the variables loaded

In [84]:
userID = 2110

In [85]:
user_id = 2 #2nd row in ratings matrix and predicted matrix

### Get the predicted ratings for userID `2110` and sort them in descending order

In [86]:
p_rating = []
p_ISBN = []
isbns = books["ISBN"].unique()
for i in isbns:
    p_rating.append(algo.predict(userID, i, verbose=False))
    p_ISBN.append(i)

### Create a dataframe with name `user_data` containing userID `2110` explicitly interacted books

In [87]:
user_data = pd.DataFrame(data = p_rating)
user_data.drop(["r_ui","details"],axis = 1, inplace = True)
user_data.columns = ["UserID","ISBN","Rating"]

In [88]:
user_data.head()

UserID        ISBN    Rating
0    2110  0195153448  8.159387
1    2110  0002005018  8.440872
2    2110  0060973129  8.169962
3    2110  0374157065  8.201638
4    2110  0393045218  8.159387

In [89]:
user_data.shape

(271357, 3)

In [90]:
user_data["ISBN"] = user_data["ISBN"].astype("object")

### Combine the user_data and and corresponding book data(`book_data`) in a single dataframe with name `user_full_info`

In [157]:
user_full_info = pd.concat([user_data,books],axis = 1,join="inner")

In [158]:
print (books.shape)
print (user_full_info.shape)

(271357, 5)
(271354, 8)


In [159]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor  yearOfPublication                   publisher  
0    Mark P. O. Morford               2002     Oxford University Press  
1  Richard Bruce Wright               2001       HarperFlamingo Canada  
2          Carlo D'Este               1991             HarperPerennial  
3      Gina Bari Kolata               1999        Farrar Straus Giroux  
4       E. J. W. Barber               1999  W. W. Norton &amp; Company

In [160]:
user_full_info.head()

UserID        ISBN    Rating        ISBN  \
0    2110  0195153448  8.159387  0195153448   
1    2110  0002005018  8.440872  0002005018   
2    2110  0060973129  8.169962  0060973129   
3    2110  0374157065  8.201638  0374157065   
4    2110  0393045218  8.159387  0393045218   

                                           bookTitle            bookAuthor  \
0                                Classical Mythology    Mark P. O. Morford   
1                                       Clara Callan  Richard Bruce Wright   
2                               Decision in Normandy          Carlo D'Este   
3  Flu: The Story of the Great Influenza Pandemic...      Gina Bari Kolata   
4                             The Mummies of Urumchi       E. J. W. Barber   

   yearOfPublication                   publisher  
0               2002     Oxford University Press  
1               2001       HarperFlamingo Canada  
2               1991             HarperPerennial  
3               1999        Farrar Straus Giroux  
4               1999  W. W. Norton &amp; Company

### Get top 10 recommendations for above given userID from the books not already rated by that user

In [161]:
user_full_info.sort_values(by="Rating",ascending = False).head(10)

UserID        ISBN     Rating        ISBN  \
563      2110  0156027321  10.000000  0156027321   
356      2110  0142001740  10.000000  0142001740   
2768     2110  0142004235  10.000000  0142004235   
6330     2110  0439136369   9.939347  0439136369   
915      2110  0440219078   9.937419  0440219078   
79370    2110  0439425220   9.903499  0439425220   
11053    2110  0446520802   9.897717  0446520802   
833      2110  0316776963   9.885665  0316776963   
12763    2110  0618002227   9.862845  0618002227   
2536     2110  0385265700   9.846360  0385265700   

                                               bookTitle        bookAuthor  \
563                                           Life of Pi       Yann Martel   
356                              The Secret Life of Bees     Sue Monk Kidd   
2768                    East of Eden (Oprah's Book Club)    John Steinbeck   
6330   Harry Potter and the Prisoner of Azkaban (Book 3)     J. K. Rowling   
915                   The Giver (21st Century Reference)        LOIS LOWRY   
79370  Harry Potter and the Chamber of Secrets Postca...     J. K. Rowling   
11053                                       The Notebook   Nicholas Sparks   
833                               Me Talk Pretty One Day     David Sedaris   
12763  The Fellowship of the Ring (The Lord of the Ri...  J. R. R. Tolkien   
2536    The Book of Ruth (Oprah's Book Club (Paperback))     Jane Hamilton   

       yearOfPublication                 publisher  
563                 2003             Harvest Books  
356                 2003             Penguin Books  
2768                2003             Penguin Books  
6330                2001                Scholastic  
915                 1994                Laure Leaf  
79370               2002                Scholastic  
11053               1996              Warner Books  
833                 2001            Back Bay Books  
12763               1999  Houghton Mifflin Company  
2536                1990                    Anchor